lang/cs/Org.Apache.REEF.Network/Elastic/Operators/Logical/ElasticOperator.cs

```diff
@@ -0,0 +1,591 @@
+            {
+                throw new IllegalStateException("Operator need to be build before gathering information.");
+            }
+
+            masterTasks.Add(Utils.BuildTaskId(Stage.StageName, MasterId));
+
+            if (_next != null)
+            {
+                _next.GatherMasterIds(ref masterTasks);
+            }
+        }
+
+        /// <summary>
+        /// Log the final statistics of the operator.
+        /// This is called when the pipeline execution is completed.
+        /// </summary>
+        internal virtual string LogFinalStatistics()
+        {
+            var str = LogInternalStatistics();
+
+            if (_next != null)
+            {
+                str += _next.LogFinalStatistics();
+            }
+
+            return str;
+        }
+
+        /// <summary>
+        /// Appends the message type to the configuration.
+        /// </summary>
+        /// <param name="operatorType">The type of the messages the operator is configured to accept</param>
+        /// <param name="confBuilder">The configuration builder the message type will be added to</param>
+        protected void SetMessageType(Type operatorType, ref ICsConfigurationBuilder confBuilder)
+        {
+            if (operatorType.IsGenericType)
+            {
+                var genericTypes = operatorType.GenericTypeArguments;
+                var msgType = genericTypes[0];
+                confBuilder.BindNamedParameter<OperatorParameters.MessageType, string>(
+                    GenericType<OperatorParameters.MessageType>.Class, msgType.AssemblyQualifiedName);
+            }
+            else
+            {
+                throw new IllegalStateException("Expecting a generic type for the message.");
+            }
+        }
+
+        /// <summary>
+        /// Action to trigger when the operator recdeives a notification that a new iteration is started.
+        /// </summary>
+        /// <param name="iteration">The new iteration number</param>
+        protected void OnNewIteration(int iteration)
+        {
+            _topology.OnNewIteration(iteration);
+
+            if (_next != null)
+            {
+                _next.OnNewIteration(iteration);
+            }
+        }
+
+        /// <summary>
+        /// This method is operator specific and serializes the operator configuration into the input list.
+        /// </summary>
+        /// <param name="serializedOperatorsConfs">A list the serialized operator configuration will be
+        /// appended to</param>
+        /// <param name="taskId">The task id of the task that belongs to this operator</param>
+        protected virtual void GetOperatorConfiguration(ref IList<string> serializedOperatorsConfs, int taskId)
+        {
+            ICsConfigurationBuilder operatorBuilder = TangFactory.GetTang().NewConfigurationBuilder();
+
+            _topology.GetTaskConfiguration(ref operatorBuilder, taskId);
+
+            PhysicalOperatorConfiguration(ref operatorBuilder);
```

Comment:
I don't quite like those `ref`s. I would much prefer immutable configurations returned from each call and then merge them here

Reply:
Fixed!